mempool/clist_mempool.go

```diff
@@ -289,51 +283,84 @@ func (mem *CListMempool) CheckTx(tx types.Tx) (*abcicli.ReqRes, error) {
-		switch checkType {
-		case abci.CHECK_TX_TYPE_CHECK:
-			if !mem.recheck.done() {
-				panic(log.NewLazySprintf("rechecking has not finished; cannot check new tx %v",
-					types.Tx(req.GetCheckTx().Tx).Hash()))
-			}
-			mem.resCbFirstTime(req.GetCheckTx().Tx, res.GetCheckTx())
+// handleCheckTxResponse handles CheckTx responses for transactions validated for the first time.
+//
+//   - txInfo optionally holds the ID of the peer that sent the transaction, if any.
+func (mem *CListMempool) handleCheckTxResponse(tx types.Tx, txInfo *TxInfo) func(res *abci.Response) {
+	return func(r *abci.Response) {
+		res := r.GetCheckTx()
+		if res == nil {
+			panic(fmt.Sprintf("unexpected response value %v not of type CheckTx", r))
+		}
 
-		case abci.CHECK_TX_TYPE_RECHECK:
-			if mem.recheck.done() {
-				mem.logger.Error("rechecking has finished; discard late recheck response",
-					"tx", log.NewLazySprintf("%v", types.Tx(req.GetCheckTx().Tx).Hash()))
-				return
-			}
-			mem.metrics.RecheckTimes.Add(1)
-			mem.resCbRecheck(req.GetCheckTx().Tx, res.GetCheckTx())
+		// Check that rechecking txs is not in process.
+		if !mem.recheck.done() {
+			panic(log.NewLazySprintf("rechecking has not finished; cannot check new tx %X", tx.Hash()))
+		}
 
-		default:
-			panic(fmt.Sprintf("unexpected value %d of RequestCheckTx.type", checkType))
+		var postCheckErr error
+		if mem.postCheck != nil {
+			postCheckErr = mem.postCheck(tx, res)
 		}
 
-		// update metrics
-		mem.metrics.Size.Set(float64(mem.Size()))
-		mem.metrics.SizeBytes.Set(float64(mem.SizeBytes()))
+		// If tx is invalid, remove it from the cache.
+		if res.Code != abci.CodeTypeOK || postCheckErr != nil {
+			mem.tryRemoveFromCache(tx)
+			mem.logger.Debug(
+				"rejected invalid transaction",
+				"tx", tx.Hash(),
+				"res", res,
+				"err", postCheckErr,
+			)
+			mem.metrics.FailedTxs.Add(1)
+			return
+		}
 
-	default:
-		// ignore other messages
+		// Check again that mempool isn't full, to reduce the chance of exceeding the limits.
+		if err := mem.isFull(len(tx)); err != nil {
+			mem.forceRemoveFromCache(tx) // mempool might have space later
+			mem.logger.Error(err.Error())
+			return
+		}
+
+		// Add tx to mempool and notify that new txs are available.
+		memTx := mempoolTx{
+			height:    mem.height.Load(),
+			gasWanted: res.GasWanted,
+			tx:        tx,
+		}
+		if mem.addTx(&memTx, txInfo) {
+			mem.notifyTxsAvailable()
+
+			// update metrics
+			mem.metrics.Size.Set(float64(mem.Size()))
+			mem.metrics.SizeBytes.Set(float64(mem.SizeBytes()))
+		}
 	}
 }
 
 // Called from:
-//   - resCbFirstTime (lock not held) if tx is valid
-func (mem *CListMempool) addTx(memTx *mempoolTx) bool {
+//   - handleCheckTxResponse (lock not held) if tx is valid
+func (mem *CListMempool) addTx(memTx *mempoolTx, txInfo *TxInfo) bool {
 	tx := memTx.tx
+	txKey := tx.Key()
+
+	// Check if the transaction is already in the mempool.
+	if elem, ok := mem.getCElement(txKey); ok {
+		if txInfo != nil {
```

Comment:
Can we have a case where `txInfo` is `nil`?

Reply:
This is a corner case, allowed by `config.KeepInvalidTxsInCache`. If we remove this option, we could simplify this code and panic if there's no sender.